### ISSUE TITLE
Spark code referring to private members fail at runtime

### ISSUE THREAD

**Issue by everson** [scala/bug#12533]
## reproduction steps

```
git clone git@github.com:everson/spark-codegen-bug.git
sbt +test
```

## problem

Starting with Scala 2.12.13, Spark code referring to private members fail at runtime. 

Tested in using JVM 8, 11 on scala versions 2.12.12 (works), 12.12.13 to 12.12.15 and 12.13.7 to 12.13.8

In a complex project, the error looks like:

```
[error]    Success(SparkFailures(NonEmpty[Unknown(org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 2.0 failed 1 times, most recent failure: Lost task 1.0 in stage 2.0 (TID 3) (192.168.0.80 executor driver): java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: Private member cannot be accessed from type "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection".
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:306)
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:293)
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
[error]    	at org.sparkproject.guava.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:135)
[error]    	at org.sparkproject.guava.cache.LocalCache$LoadingValueReference.waitForValue(LocalCache.java:3620)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.waitForLoadingValue(LocalCache.java:2362)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2349)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.get(LocalCache.java:2257)
[error]    	at org.sparkproject.guava.cache.LocalCache.get(LocalCache.java:4000)
[error]    	at org.sparkproject.guava.cache.LocalCache.getOrLoad(LocalCache.java:4004)
[error]    	at org.sparkproject.guava.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4874)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$.compile(CodeGenerator.scala:1351)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:205)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:39)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator.generate(CodeGenerator.scala:1277)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator.generate(CodeGenerator.scala:1274)
[error]    	at org.apache.spark.sql.execution.ObjectOperator$.deserializeRowToObject(objects.scala:147)
[error]    	at org.apache.spark.sql.execution.AppendColumnsExec.$anonfun$doExecute$12(objects.scala:326)
[error]    	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
[error]    	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
[error]    	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
[error]    	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
[error]    	at org.apache.spark.scheduler.Task.run(Task.scala:131)
[error]    	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
[error]    	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
[error]    	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
[error]    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]    	at java.lang.Thread.run(Thread.java:748)
[error]    Caused by: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: Private member cannot be accessed from type "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection".
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$.org$apache$spark$sql$catalyst$expressions$codegen$CodeGenerator$$doCompile(CodeGenerator.scala:1415)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$$anon$1.load(CodeGenerator.scala:1500)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$$anon$1.load(CodeGenerator.scala:1497)
[error]    	at org.sparkproject.guava.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3599)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.loadSync(LocalCache.java:2379)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2342)
[error]    	... 32 more
[error]
[error]    Driver stacktrace:)])) is a Success but SparkFailures(NonEmpty[Unknown(org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 2.0 failed 1 times, most recent failure: Lost task 1.0 in stage 2.0 (TID 3) (192.168.0.80 executor driver): java.util.concurrent.ExecutionException: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: Private member cannot be accessed from type "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection".
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:306)
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:293)
[error]    	at org.sparkproject.guava.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
[error]    	at org.sparkproject.guava.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:135)
[error]    	at org.sparkproject.guava.cache.LocalCache$LoadingValueReference.waitForValue(LocalCache.java:3620)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.waitForLoadingValue(LocalCache.java:2362)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2349)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.get(LocalCache.java:2257)
[error]    	at org.sparkproject.guava.cache.LocalCache.get(LocalCache.java:4000)
[error]    	at org.sparkproject.guava.cache.LocalCache.getOrLoad(LocalCache.java:4004)
[error]    	at org.sparkproject.guava.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4874)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$.compile(CodeGenerator.scala:1351)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:205)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.GenerateSafeProjection$.create(GenerateSafeProjection.scala:39)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator.generate(CodeGenerator.scala:1277)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator.generate(CodeGenerator.scala:1274)
[error]    	at org.apache.spark.sql.execution.ObjectOperator$.deserializeRowToObject(objects.scala:147)
[error]    	at org.apache.spark.sql.execution.AppendColumnsExec.$anonfun$doExecute$12(objects.scala:326)
[error]    	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:898)
[error]    	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:898)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
[error]    	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:373)
[error]    	at org.apache.spark.rdd.RDD.iterator(RDD.scala:337)
[error]    	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
[error]    	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
[error]    	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
[error]    	at org.apache.spark.scheduler.Task.run(Task.scala:131)
[error]    	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
[error]    	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
[error]    	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
[error]    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error]    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error]    	at java.lang.Thread.run(Thread.java:748)
[error]    Caused by: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 63, Column 8: Private member cannot be accessed from type "org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificSafeProjection".
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$.org$apache$spark$sql$catalyst$expressions$codegen$CodeGenerator$$doCompile(CodeGenerator.scala:1415)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$$anon$1.load(CodeGenerator.scala:1500)
[error]    	at org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator$$anon$1.load(CodeGenerator.scala:1497)
[error]    	at org.sparkproject.guava.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3599)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.loadSync(LocalCache.java:2379)
[error]    	at org.sparkproject.guava.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2342)
[error]    	... 32 more
[error]
[error]    Driver stacktrace:)]) != Success (Expectations.scala:57)
```

**Comment by som-snytt** [comment]
```
InnerClasses:
<   public static #12= #2 of #11;           // PrivNumBool=class everson/sparkcodegen/FunctionsSpec$PrivNumBool of class everson/sparkcodegen/FunctionsSpec
---
>   private static #12= #2 of #11;          // PrivNumBool=class everson/sparkcodegen/FunctionsSpec$PrivNumBool of class everson/sparkcodegen/FunctionsSpec
```
perhaps https://github.com/scala/scala/pull/9131

**Comment by SethTisue** [comment]
is the problem reproducible in Scala 2.12.15 and/or 2.13.8? (if @som-snytt's link is right, presumably it isn't.)

**Comment by lrytz** [comment]
Without looking in detail, it's possible that Spark's code generator incorrectly assumes that the inner class (https://github.com/everson/spark-codegen-bug/blob/main/src/test/scala/everson/sparkcodegen/FunctionsSpec.scala#L45) would be public. Could you get us the content of the `generated.java` file?

**Comment by som-snytt** [comment]
> presumably it isn't.

@SethTisue typo alert

The linked ticket on the linked ticket explains that the attribute is for Java language semantics, so it matters if you're generating Java code.

**Comment by everson** [comment]
@lrytz Sorry, I was unable to get a representation of `generated.java` yet. I will try more after work.

> Could you get us the content of the generated.java file?

I tried https://stackoverflow.com/a/56460270 and it crashes even before printing the generated code.

> it's possible that Spark's code generator incorrectly assumes that the inner class would be public

I am thinking Spark codegen assumes inner classes to be public. If that is the case, given it worked up to 2.12.12, how would this be handled? A bug created against Spark project itself?

**Comment by som-snytt** [comment]
Google says I can update your example code
```
Logger.getLogger("org.apache.spark.sql.catalyst.expressions.codegen.CodeGenerator").setLevel(Level.DEBUG)
```
to see
```
/* 031 */     boolean isNull_3 = i.isNullAt(0);
/* 032 */     everson.sparkcodegen.FunctionsSpec$PrivNumBool value_3 = isNull_3 ?
/* 033 */     null : ((everson.sparkcodegen.FunctionsSpec$PrivNumBool)i.get(0, null));
/* 034 */     if (isNull_3) {
/* 035 */       throw new NullPointerException(((java.lang.String) references[0] /* errMsg */));
/* 036 */     }
/* 037 */     boolean isNull_0 = true;
/* 038 */     int value_0 = -1;
/* 039 */     isNull_0 = false;
/* 040 */     if (!isNull_0) {
/* 041 */       value_0 = value_3.someNum();
/* 042 */     }
[snip]
22/02/08 13:19:26 ERROR CodeGenerator: failed to compile: org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 41, Column 19: No applicable constructor/method found for zero actual parameters; candidates are: "public int everson.sparkcodegen.FunctionsSpec$PrivNumBool.someNum()"
org.codehaus.commons.compiler.CompileException: File 'generated.java', Line 41, Column 19: No applicable constructor/method found for zero actual parameters; candidates are: "public int everson.sparkcodegen.FunctionsSpec$PrivNumBool.someNum()"
        at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:12021)
        at org.codehaus.janino.UnitCompiler.findMostSpecificIInvocable(UnitCompiler.java:9332)
        at org.codehaus.janino.UnitCompiler.findIMethod(UnitCompiler.java:9192)
        at org.codehaus.janino.UnitCompiler.findIMethod(UnitCompiler.java:9110)
        at org.codehaus.janino.UnitCompiler.compileGet2(UnitCompiler.java:5055)
        at org.codehaus.janino.UnitCompiler.access$9100(UnitCompiler.java:226)
```
I confirmed it's the same generated code for both Scala versions.

I guess that is the correct error for the example, although it is different from the error observed in the field.

BTW example requires old JDK for options and module access.

all LOC
```
22/02/08 13:19:26 DEBUG CodeGenerator:
/* 001 */ public java.lang.Object generate(Object[] references) {
/* 002 */   return new SpecificUnsafeProjection(references);
/* 003 */ }
/* 004 */
/* 005 */ class SpecificUnsafeProjection extends org.apache.spark.sql.catalyst.expressions.UnsafeProjection {
/* 006 */
/* 007 */   private Object[] references;
/* 008 */   private org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[] mutableStateArray_0 = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter[1];
/* 009 */
/* 010 */   public SpecificUnsafeProjection(Object[] references) {
/* 011 */     this.references = references;
/* 012 */     mutableStateArray_0[0] = new org.apache.spark.sql.catalyst.expressions.codegen.UnsafeRowWriter(2, 0);
/* 013 */
/* 014 */   }
/* 015 */
/* 016 */   public void initialize(int partitionIndex) {
/* 017 */
/* 018 */   }
/* 019 */
/* 020 */   // Scala.Function1 need this
/* 021 */   public java.lang.Object apply(java.lang.Object row) {
/* 022 */     return apply((InternalRow) row);
/* 023 */   }
/* 024 */
/* 025 */   public UnsafeRow apply(InternalRow i) {
/* 026 */     mutableStateArray_0[0].reset();
/* 027 */
/* 028 */
/* 029 */     mutableStateArray_0[0].zeroOutNullBytes();
/* 030 */
/* 031 */     boolean isNull_3 = i.isNullAt(0);
/* 032 */     everson.sparkcodegen.FunctionsSpec$PrivNumBool value_3 = isNull_3 ?
/* 033 */     null : ((everson.sparkcodegen.FunctionsSpec$PrivNumBool)i.get(0, null));
/* 034 */     if (isNull_3) {
/* 035 */       throw new NullPointerException(((java.lang.String) references[0] /* errMsg */));
/* 036 */     }
/* 037 */     boolean isNull_0 = true;
/* 038 */     int value_0 = -1;
/* 039 */     isNull_0 = false;
/* 040 */     if (!isNull_0) {
/* 041 */       value_0 = value_3.someNum();
/* 042 */     }
/* 043 */     mutableStateArray_0[0].write(0, value_0);
/* 044 */
/* 045 */     boolean isNull_7 = i.isNullAt(0);
/* 046 */     everson.sparkcodegen.FunctionsSpec$PrivNumBool value_7 = isNull_7 ?
/* 047 */     null : ((everson.sparkcodegen.FunctionsSpec$PrivNumBool)i.get(0, null));
/* 048 */     if (isNull_7) {
/* 049 */       throw new NullPointerException(((java.lang.String) references[1] /* errMsg */));
/* 050 */     }
/* 051 */     boolean isNull_4 = true;
/* 052 */     boolean value_4 = false;
/* 053 */     isNull_4 = false;
/* 054 */     if (!isNull_4) {
/* 055 */       value_4 = value_7.expected();
/* 056 */     }
/* 057 */     mutableStateArray_0[0].write(1, value_4);
/* 058 */     return (mutableStateArray_0[0].getRow());
/* 059 */   }
/* 060 */
/* 061 */
/* 062 */ }
```

**Comment by lrytz** [comment]
Thanks for digging. Not sure why the error is differnet, why there's no error reported at `FunctionsSpec$PrivNumBool` in your case. But it seems quite sure that this should be a ticket for the Spark code generator. Maybe they cannot work around it, but make a better message bubble up (saying that `PrivNumBool` is not accessible).

**Comment by everson** [comment]
Thank you all for looking into this. I will open a ticket in the Spark project.

> som-snytt: Google says I can update your example code

Funny thing is that I saw a few answers on how to do it and that one seemed too simple to work (I thought it required I setting up logger config first). Lesson learned. 

 > Maybe they cannot work around it, but make a better message bubble up (saying that PrivNumBool is not accessible).

Right, even in minimal case, the error message is not very helpful. In the project I was working on when I found this, it was super hard to isolate the issue.
